File: android/app/src/main/java/com/agentmobile/agent/UdpMedia.java
package com.agentmobile.agent;

import android.os.SystemClock;

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;

/**
 * UdpMedia — Android side of the UDP media path (transport/udp-media.js).
 *
 * Audio rides UDP instead of WebSocket/TCP for spotty cellular. Security stays on
 * the SAME session {@link KoCrypto.Channel}: each datagram is one sealed frame
 * `[type][nonce][tag][ct]` on nonce stream {@link KoCrypto#STREAM_UDP} (its own
 * counter + anti-replay window, independent of the WS stream), matching the
 * sidecar's UdpMedia exactly.
 *
 * Authenticated plaintext per datagram: `[kind u8][seq u32][tsMs u64][opus]`
 *   kind 0 = mic uplink (this device -> gateway), kind 1 = reply downlink,
 *   kind 2 = probe/ack. On start this class sends kind-2 probes every second
 *   until the sidecar answers with an authenticated kind-2 ack; only then does
 *   {@link #onProbed(Callback)} fire so the host can flip the uplink to UDP.
 *
 * The receiver funnels downlink frames through an {@link Ajb} so playout is
 * ordered + gap-concealed (PLC markers), then hands each frame to `sink`.
 * Send + receive are on separate threads; the jitter buffer is synchronized.
 */
public final class UdpMedia {
    public interface Sink {
        /** kind, seq, tsMs, opus (null when concealed), concealed */
        void onMedia(int kind, long seq, long tsMs, byte[] opus, boolean concealed);
    }
    public interface Callback { void run(); }

    private static final int PROBE_INTERVAL_MS = 1000;      // while not yet acked
    private static final int KEEPALIVE_INTERVAL_MS = 5000;  // once up: liveness probes
    private static final int DOWN_AFTER_MS = 16000;         // no ack for this long -> path DOWN

    private final KoCrypto.Channel channel;
    private final Sink sink;
    private final Ajb jitter = new Ajb(6, 60, 20, 80, 100);

    private volatile DatagramSocket socket;
    private volatile boolean running;
    private volatile boolean probed;
    private volatile long lastAckMs;
    private Thread rxThread;
    private Thread probeThread;
    private InetAddress gwAddr;
    private int gwPort;
    private volatile Callback probedCb;
    private volatile Callback downCb;

    public UdpMedia(KoCrypto.Channel channel, Sink sink) {
        this.channel = channel; this.sink = sink;
    }

    /** Fired (on the udp-rx thread) each time the path comes UP: the sidecar's authenticated
     *  kind-2 ack arrives after a probe (first time, or again after a DOWN). */
    public void onProbed(Callback cb) { this.probedCb = cb; }
    /** Fired (on the probe thread) when acks stop for DOWN_AFTER_MS (NAT rebind, network
     *  switch, sidecar restart): the host must fall back to the WS uplink. Probing continues
     *  and onProbed fires again when the path recovers. */
    public void onDown(Callback cb) { this.downCb = cb; }
    public boolean isProbed() { return probed; }

    /** Bind the local socket. Returns the chosen local port (the gateway learns
     * it from our first uplink datagram and replies to it). */
    public int start() throws Exception {
        socket = new DatagramSocket(); // ephemeral local port on all interfaces
        running = true;
        rxThread = new Thread(this::receiveLoop, "agent-udp-rx");
        rxThread.setDaemon(true);
        rxThread.start();
        probeThread = new Thread(this::probeLoop, "agent-udp-probe");
        probeThread.setDaemon(true);
        probeThread.start();
        return socket.getLocalPort();
    }

    private void probeLoop() {
        int pseq = 0;
        long nextProbe = 0;
        while (running) {
            long now = SystemClock.elapsedRealtime();
            if (probed && now - lastAckMs > DOWN_AFTER_MS) {
                probed = false;                         // path went dark: WS fallback
                Callback cb = downCb;
                if (cb != null) { try { cb.run(); } catch (Exception ignored) {} }
            }
            if (now >= nextProbe) {
                sendFrame(2, pseq++, System.currentTimeMillis(), new byte[0]);
                nextProbe = now + (probed ? KEEPALIVE_INTERVAL_MS : PROBE_INTERVAL_MS);
            }
            jitter.tick();                              // 1 Hz: lets the loss penalty decay
            try { Thread.sleep(1000); } catch (InterruptedException e) { break; }
        }
    }

    /** Set the gateway media endpoint (host from the WS URL, port from hello). */
    public void setPeer(InetAddress addr, int port) { this.gwAddr = addr; this.gwPort = port; }

    public boolean isUp() { return running && gwAddr != null; }

    /** Seal + send one media frame (called from the mic encode thread). */
    public boolean sendFrame(int kind, long seq, long tsMs, byte[] opus) {
        DatagramSocket s = socket;
        if (s == null || gwAddr == null) return false;
        try {
            byte[] pt = new byte[13 + opus.length];
            pt[0] = (byte) kind;
            byte[] seqB = {(byte)(seq >> 24), (byte)(seq >> 16), (byte)(seq >> 8), (byte)seq};
            System.arraycopy(seqB, 0, pt, 1, 4);
            for (int i = 0; i < 8; i++) pt[5 + i] = (byte) (tsMs >>> (56 - i * 8));
            System.arraycopy(opus, 0, pt, 13, opus.length);
            byte[] wire = channel.seal(KoCrypto.TYPE_AUDIO, pt, KoCrypto.STREAM_UDP);
            s.send(new DatagramPacket(wire, wire.length, gwAddr, gwPort));
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    /** Any authenticated downlink datagram also proves the path is alive. */
    private void touchAlive() { lastAckMs = SystemClock.elapsedRealtime(); }

    private void receiveLoop() {
        byte[] bufArr = new byte[4096];
        DatagramPacket pkt = new DatagramPacket(bufArr, bufArr.length);
        while (running) {
            try {
                pkt.setLength(bufArr.length);
                socket.receive(pkt);
                byte[] wire = java.util.Arrays.copyOfRange(pkt.getData(), 0, pkt.getLength());
                byte[] pt;
                try { pt = channel.open(wire); }          // AEAD + anti-replay; throws -> drop
                catch (Exception bad) { continue; }
                if (pt == null || pt.length < 13) continue;
                int kind = pt[0] & 0xff;
                long seq = ((pt[1] & 0xffL) << 24) | ((pt[2] & 0xffL) << 16) | ((pt[3] & 0xffL) << 8) | (pt[4] & 0xffL);
                long tsMs = 0; for (int i = 0; i < 8; i++) tsMs = (tsMs << 8) | (pt[5 + i] & 0xffL);
                if (kind == 2) {                // sidecar ack to our probe — UDP is friendly now
                    lastAckMs = SystemClock.elapsedRealtime();
                    if (!probed) {
                        probed = true;
                        Callback cb = probedCb;
                        if (cb != null) { try { cb.run(); } catch (Exception ignored) {} }
                    }
                    continue;
                }
                touchAlive();
                byte[] opus = java.util.Arrays.copyOfRange(pt, 13, pt.length);
                jitter.push(seq, tsMs, opus);
                drain(tsMs);
            } catch (Exception e) {
                if (!running) break;
            }
        }
    }

    private void drain(long tsMs) {
        Ajb.Frame f;
        long now = SystemClock.elapsedRealtime();
        while ((f = jitter.pull(now)) != null) {
            sink.onMedia(1, f.seq, f.concealed ? tsMs : f.tsMs, f.opus, f.concealed);
        }
    }

    public void close() {
        running = false;
        DatagramSocket s = socket;
        if (s != null) { try { s.close(); } catch (Exception ignored) {} }
    }
}

File: transport/udp-media.js
// udp-media.js — UDP media transceiver for the spotty-cellular transport path.
//
// Audio moves off WebSocket/TCP onto a UDP RTP-style socket (see
// docs/udp-media-transport.md). Security stays on the EXISTING ChaCha20-Poly1305
// proto Channel — each UDP datagram is one authenticated box; the nonce rides in
// the frame. UDP uses its OWN nonce stream (stream 1) with a per-stream counter
// and anti-replay window, so WS stalls and UDP reordering never interfere.
//
// Wire layout per UDP datagram (authenticated plaintext):
//   [ kind u8 ][ seq u32 ][ tsMs u64 ][ opus ... ]
//   kind: 0 = mic uplink (phone->gateway), 1 = reply downlink (gateway->phone)
//        2 = probe/ack: phone sends kind-2 probes until it receives an
//            authenticated kind-2 ack (echoing the probe seq) — only then does
//            it flip its uplink to UDP. The ack proves Bidirectional reach.
//
// The receiver funnels frames through the AdaptiveJitterBuffer so playout is
// ordered, gap-concealed (PLC), and always pre-buffered — the reliability core.
//
// API
//   const m = new UdpMedia({ channel, allow: (rinfo)=>bool });
//   await m.start(port, host);       // binds, resolves m.port
//   m.on('media', ({kind, seq, tsMs, frame, concealed}) => …)
//   m.send({ kind, seq, tsMs, opus, addr, port })   // seals + sends
//   m.close();
//
// `concealed:true` marks a frame the sender lost — the consumer runs Opus PLC,
// never silence. `frame` is the raw opus bytes for real frames, null otherwise.
import { createSocket } from 'node:dgram';
import { unpack, T } from './wsframes.js';
import { STREAM_UDP } from '../proto.js';

export class UdpMedia {
  constructor({ channel, allow = () => true, respondToProbes = true }) {
    this.channel = channel;               // proto.Channel (AEAD)
    this.allow = allow;                   // authz: who may reach this media port
    this.respondToProbes = respondToProbes; // gateway acks kind-2 probes; phone (Java) does not
    this.socket = createSocket('udp4');
    this.port = 0;
    this.jb = null;                       // adaptive jitter buffer (created on first frame)
    this._handlers = new Map();
    this._next = null;                    // first seq seen -> stream anchor
    this._firstReq = null;                // most recent AUTHENTICATED sender (rinfo) for replies
    this.lastKind = 0;                    // kind of the most recent frame (diagnostics)
    this.lastTsMs = 0;
    this._tick = null;
  }

  on(evt, fn) { (this._handlers.get(evt) || this._handlers.set(evt, []).get(evt)).push(fn); }
  _emit(evt, ...args) { for (const fn of this._handlers.get(evt) || []) { try { fn(...args); } catch (e) { /* swallow */ } } }

  start({ port = 0, host = '0.0.0.0' } = {}) {
    return new Promise((resolve, reject) => {
      this.socket.on('error', reject);
      this.socket.on('message', (msg, rinfo) => {
        try { this._onDatagram(msg, rinfo); } catch (e) { this._emit('error', e); }
      });
      this.socket.bind(port, host, () => {
        this.port = this.socket.address().port;
        this._tick = setInterval(() => { if (this.jb) this.jb.tick(); }, 1000);
        if (this._tick.unref) this._tick.unref();
        resolve(this);
      });
    });
  }

  _onDatagram(msg, rinfo) {
    if (!this.allow(rinfo)) return;
    // Nothing below may throw out of the dgram handler: an unauthenticated peer
    // sending one junk datagram must not be able to crash the receiver.
    let pt = null;
    try {
      const f = unpack(msg);
      pt = this.channel.recvBytes(f);       // AEAD + per-stream anti-replay (stream 1 = UDP)
    } catch { pt = null; }
    if (pt === null) { this._emit('authfail', rinfo); return; }   // AEAD is the real gate
    if (pt.length < 13) return;
    const kind = pt[0];
    const seq = pt.readUInt32BE(1);
    const tsMs = Number(pt.readBigUInt64BE(5));
    const opus = Buffer.from(pt.subarray(13));
    // Learn / refresh the reply peer from the newest FRESH authenticated datagram
    // (replayed frames never get here, so a captured packet from a spoofed source
    // cannot redirect the downlink; a NAT rebind or network switch is followed).
    const prev = this._firstReq;
    if (!prev || prev.address !== rinfo.address || prev.port !== rinfo.port) {
      this._firstReq = rinfo; this._emit('peer', rinfo);
    }

    // Probe (kind 2, phone->sidecar)? Ack it and do NOT feed the jitter buffer.
    // The phone flips to UDP uplink only after it sees this authenticated ack.
    if (kind === 2) {
      if (this.respondToProbes) this.ack({ to: rinfo, seq });
      this._emit('probe', { seq, from: rinfo });
      return;
    }

    if (this.jb === null) this.jb = new AdaptiveJitterBuffer({ frameMs: 20 }); // TODO: per-stream
    this.lastKind = kind;
    this.lastTsMs = tsMs;
    this.jb.push(seq, tsMs, { kind, opus });   // frame carries ITS OWN kind (not the latest one's)
    this.pump();
    this._emit('rcv', { kind, seq, tsMs, from: rinfo });
  }

  // Authenticated kind-2 ack back to the sender. Echoes the probe seq so the
  // phone can correlate; genuine-ness is guaranteed by AEAD (only a peer with
  // the session key can mint a box that decrypts here and back on the phone).
  ack({ to, seq, kind = 2 }) {
    const body = Buffer.alloc(13);
    body[0] = kind;
    body.writeUInt32BE(seq, 1);
    body.writeBigUInt64BE(BigInt(Date.now()), 5);
    const box = this.channel.send(body, T.audio, STREAM_UDP);
    const wire = Buffer.alloc(1 + box.nonce.length + box.tag.length + box.ct.length);
    wire[0] = T.audio;
    box.nonce.copy(wire, 1);
    box.tag.copy(wire, 1 + 12);
    box.ct.copy(wire, 1 + 12 + 16);
    this.socket.send(wire, to.port, to.address);
    return true;
  }

  // Pump the jitter buffer: emit ordered frames (or PLC markers) to consumers.
  // `nowMs` injectable so tests can drive the playout clock deterministically.
  pump(nowMs = Date.now()) {
    if (!this.jb) return;
    let out;
    while ((out = this.jb.pull(nowMs))) {
      const f = out.frame;   // { kind, opus } or null (concealed)
      this._emit('media', { kind: f ? f.kind : this.lastKind, seq: out.seq, tsMs: out.tsMs,
                            frame: f ? f.opus : null, concealed: out.concealed });
    }
  }

  // Seal and send one media frame over UDP.
  send({ kind, seq, tsMs, opus, addr, port }) {
    if (!addr || !port) return false;
    const body = Buffer.alloc(13 + opus.length);
    body[0] = kind;
    body.writeUInt32BE(seq, 1);
    body.writeBigUInt64BE(BigInt(tsMs), 5);
    opus.copy(body, 13);
    const box = this.channel.send(body, T.audio, STREAM_UDP); // UDP stream counter, type as AAD
    const wire = Buffer.alloc(1 + box.nonce.length + box.tag.length + box.ct.length);
    wire[0] = T.audio;
    box.nonce.copy(wire, 1);
    box.tag.copy(wire, 1 + 12);
    box.ct.copy(wire, 1 + 12 + 16);
    this.socket.send(wire, port, addr);
    return true;
  }

  get learnedPeer() { return this._firstReq; }
  close() { if (this._tick) clearInterval(this._tick); try { this.socket.close(); } catch {} }
}

import { AdaptiveJitterBuffer } from './adaptive-jitter.js';
